README.md
# Cinerino LINE Assistant

[![CircleCI](https://circleci.com/gh/cinerino/line-assistant.svg?style=svg)](https://circleci.com/gh/cinerino/line-assistant)

## Table of contents

- [Cinerino LINE Assistant](#cinerino-line-assistant)
  - [Table of contents](#table-of-contents)
  - [Usage](#usage)
    - [Environment variables](#environment-variables)
  - [Code Samples](#code-samples)
  - [Jsdoc](#jsdoc)
  - [License](#license)
  - [Reference](#reference)
    - [LINE Reference](#line-reference)

## Usage

### Environment variables

| Name                               | Required | Purpose                   | Value                                                                 |
| ---------------------------------- | -------- | ------------------------- | --------------------------------------------------------------------- |
| `DEBUG`                            | false    | cinerino-line-assistant:* | Debug                                                                 |
| `NODE_ENV`                         | true     |                           | environment name                                                      |
| `SENDGRID_API_KEY`                 | true     |                           | SendGrid API Key                                                      |
| `GMO_ENDPOINT`                     | true     |                           | GMO API endpoint                                                      |
| `GMO_SITE_ID`                      | true     |                           | GMO SiteID                                                            |
| `GMO_SITE_PASS`                    | true     |                           | GMO SitePass                                                          |
| `COA_ENDPOINT`                     | true     |                           | COA API endpoint                                                      |
| `COA_REFRESH_TOKEN`                | true     |                           | COA API refresh token                                                 |
| `AZURE_STORAGE_CONNECTION_STRING`  | true     |                           | Save CSV files on azure storage                                       |
| `LINE_BOT_CHANNEL_SECRET`          | true     |                           | LINE Messaging API 署名検証                                           |
| `LINE_BOT_CHANNEL_ACCESS_TOKEN`    | true     |                           | LINE Messaging API 認証                                               |
| `API_AUTHORIZE_SERVER_DOMAIN`      | true     |                           | API 認可サーバードメイン                                              |
| `API_CLIENT_ID`                    | true     |                           | APIクライアントID                                                     |
| `API_CLIENT_SECRET`                | true     |                           | APIクライアントシークレット                                           |
| `API_TOKEN_ISSUER`                 | true     |                           | APIトークン発行者                                                     |
| `API_CODE_VERIFIER`                | true     |                           | API認可コード検証鍵                                                   |
| `USER_REFRESH_TOKEN`               | false    |                           | APIのリフレッシュトークン(セットすると認証をスキップできる、開発用途) |
| `REDIS_HOST`                       | true     |                           | ログイン状態保持ストレージ                                            |
| `REDIS_PORT`                       | true     |                           | ログイン状態保持ストレージ                                            |
| `REDIS_KEY`                        | true     |                           | ログイン状態保持ストレージ                                            |
| `USER_EXPIRES_IN_SECONDS`          | true     |                           | ユーザーセッション保持期間                                            |
| `REFRESH_TOKEN_EXPIRES_IN_SECONDS` | true     |                           | リフレッシュトークン保管期間                                          |
| `AWS_ACCESS_KEY_ID`                | true     |                           |                                                                       |
| `AWS_SECRET_ACCESS_KEY`            | true     |                           |                                                                       |
| `FACE_MATCH_THRESHOLD`             | true     |                           | 顔認証閾値                                                            |
| `CINERINO_CONSOLE_ENDPOINT`        | true     |                           |                                                                       |
| `PROJECT_ID`                       | true     |                           |                                                                       |

## Code Samples

Code sample are [here](https://github.com/cinerino/line-assistant/tree/master/example).

## Jsdoc

`npm run doc` emits jsdoc to ./doc.

## License

ISC

## Reference

### LINE Reference

* [LINE BUSSINESS CENTER](https://business.line.me/ja/)
* [LINE@MANAGER](https://admin-official.line.me/)
* [API Reference](https://devdocs.line.me/ja/)
* [LINE Pay技術サポート](https://pay.line.me/jp/developers/documentation/download/tech?locale=ja_JP)
* [LINE Pay Home](https://pay.line.me/jp/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## Unreleased

### Added

### Changed

- update @cinerino/api-nodejs-client
- Redisコネクション処理調整

### Deprecated

### Removed

### Fixed

### Security

## v3.0.0 - 2019-03-10

### Changed

- @cinerino/api-nodejs-clientで再構築

## v2.5.2 - 2019-01-30

### Changed

- install @motionpicture/sskts-api-nodejs-client@5.0.0
- install @motionpicture/sskts-domain@27.0.0

## v2.5.1 - 2018-12-10

### Changed

- update sskts-domain

## v2.5.0 - 2018-06-10

### Changed

- 取引照会結果を複数注文アイテムタイプに対応。
- 取引レポートの内容を拡張。
- install sskts-domain@25.x.x

## v2.4.1 - 2018-04-19

### Fixed

- 期限切れ取引の承認アクション履歴の取得方法を修正。

## v2.4.0 - 2018-03-28

### Added

- Face Loginを追加。

## v2.3.2 - 2018-03-27

### Fixed

- 取引照会結果の文字数が多い場合の対応。

## v2.3.1 - 2018-02-26
- update sskts-domain.

## v2.3.0 - 2018-02-20
### Added
- 注文返品機能を追加。
- 注文照会にアクション履歴情報を追加。
- 取引IDによる照会機能を追加。期限入れの取引詳細も確認できるように調整。
- 注文照会に注文アイテム状態情報を追加。

## v2.2.0 - 2018-02-05
### Changed
- Cognito管理者ユーザーとの連携を調整。
- ログアウト機能を追加。

## v2.1.2 - 2017-12-07
### Added
- ユーザーセッション保持期間設定を追加。

### Changed
- ユーザーの状態にウェブフックイベント自体を引き継ぐように変更。

## v2.1.1 - 2017-12-07
### Changed
- サインイン後のページに戻るリンクを追加。

## v2.1.0 - 2017-12-06
### Added
- Cognitoユーザー認証追加。

### Changed
- install [@motionpicture/sskts-domain@23.4.0](https://www.npmjs.com/package/@motionpicture/sskts-domain)
- 取引csv検索条件を調整。

## v2.0.0 - 2017-10-31
### Changed
- sskts-domain@v23(スキーマ一新)に対応。

## v1.3.0 - 2017-07-04
### Added
- 予約番号検索を、成立以外の取引にも対応。
- 処理中のステータスをメッセージとして送信するように対応。

### Changed
- タスク仕様に合わせて調整。

## v1.2.0 - 2017-06-26
### Changed
- 予約照会条件に劇場コード追加

### Security
- [typescript@^2.4.0](https://github.com/Microsoft/TypeScript)

## v1.1.5 - 2017-06-12
### Added
- 取引詳細にGMOショップ情報を追加。

### Changed
- 取引詳細テンプレート調整。

## v1.1.4 - 2017-06-01
### Added
- csvの項目に劇場コードと予約番号を追加。

### Fixed
- csvの検索日時toが結果に含まれないバグを修正。

## v1.1.3 - 2017-05-31
### Fixed
- 多言語スキーマを追加したことに対する互換性対応。

## v1.1.2 - 2017-05-17
### Fixed
- csvダウンロードにて、ムビチケ購入管理番号表示が崩れるバグを修正。

## v1.1.1 - 2017-05-17
### Fixed
- 座席予約資産の券種名取得方法を修正。

## v1.1.0 - 2017-05-16
### Added
- 取引csvダウンロード機能を追加。

## v1.0.12 - 2017-04-19
### Changed
- QRコード表示をリンクに変更。

## v1.0.11 - 2017-04-19
### Added
- 電話番号検索を実験的に追加。

## v1.0.5 - 2017-04-18
### Added
- ムビチケ座席予約の場合に対応。

## v1.0.4 - 2017-04-18
### Changed
- 予約番号照会の取引内容を調整。

## v1.0.2 - 2017-04-14
### Changed
- パッケージ依存関係を整理。

## v1.0.1 - 2017-04-14
### Added
- 取引検索結果の情報を強化。

## v1.0.0 - 2017-04-01
### Added
- initialize


dst/app/user.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const cinerinoapi = require("@cinerino/api-nodejs-client");
const AWS = require("aws-sdk");
const createDebug = require("debug");
const redis = require("ioredis");
const jwt = require("jsonwebtoken");
const debug = createDebug('cinerino-line-assistant:user');
// 以下環境変数をセットすること
// AWS_ACCESS_KEY_ID
// AWS_SECRET_ACCESS_KEY
const rekognition = new AWS.Rekognition({
    apiVersion: '2016-06-27',
    region: 'us-west-2'
});
const redisClient = new redis({
    host: process.env.REDIS_HOST,
    port: Number(process.env.REDIS_PORT),
    password: process.env.REDIS_KEY,
    tls: (process.env.REDIS_TLS_SERVERNAME !== undefined) ? { servername: process.env.REDIS_TLS_SERVERNAME } : undefined
});
const USER_EXPIRES_IN_SECONDS = process.env.USER_EXPIRES_IN_SECONDS;
if (USER_EXPIRES_IN_SECONDS === undefined) {
    throw new Error('Environment variable USER_EXPIRES_IN_SECONDS required.');
}
// tslint:disable-next-line:no-magic-numbers
const EXPIRES_IN_SECONDS = parseInt(USER_EXPIRES_IN_SECONDS, 10);
const POST_EVENT_TOKEN_EXPIRES_IN_SECONDS = 60;
const SECRET = 'secret';
const REFRESH_TOKEN_EXPIRES_IN_SECONDS_ENV = process.env.REFRESH_TOKEN_EXPIRES_IN_SECONDS;
if (REFRESH_TOKEN_EXPIRES_IN_SECONDS_ENV === undefined) {
    throw new Error('Environment variable REFRESH_TOKEN_EXPIRES_IN_SECONDS required.');
}
// tslint:disable-next-line:no-magic-numbers
const REFRESH_TOKEN_EXPIRES_IN_SECONDS = parseInt(REFRESH_TOKEN_EXPIRES_IN_SECONDS_ENV, 10);
/**
 * LINEユーザー
 * @see https://aws.amazon.com/blogs/mobile/integrating-amazon-cognito-user-pools-with-api-gateway/
 */
class User {
    constructor(configurations) {
        this.host = configurations.host;
        this.userId = configurations.userId;
        this.state = configurations.state;
        this.rekognitionCollectionId = `cinerino-line-assistant-${this.userId}`;
        this.authClient = new cinerinoapi.auth.OAuth2({
            domain: process.env.API_AUTHORIZE_SERVER_DOMAIN,
            clientId: process.env.API_CLIENT_ID,
            clientSecret: process.env.API_CLIENT_SECRET,
            redirectUri: `https://${configurations.host}/signIn`,
            logoutUri: `https://${configurations.host}/logout`
        });
    }
    generateAuthUrl() {
        return this.authClient.generateAuthUrl({
            scopes: [],
            state: this.state,
            codeVerifier: process.env.API_CODE_VERIFIER
        });
    }
    generateLogoutUrl() {
        return this.authClient.generateLogoutUrl();
    }
    getCredentials() {
        return __awaiter(this, void 0, void 0, function* () {
            return redisClient.get(`line-assistant.credentials.${this.userId}`)
                .then((value) => (value === null) ? null : JSON.parse(value));
        });
    }
    getRefreshToken() {
        return __awaiter(this, void 0, void 0, function* () {
            return redisClient.get(`line-assistant.refreshToken.${this.userId}`)
                .then((value) => (value === null) ? null : value);
        });
    }
    setCredentials(credentials) {
        const payload = jwt.decode(credentials.access_token);
        debug('payload:', payload);
        this.payload = payload;
        this.accessToken = credentials.access_token;
        this.authClient.setCredentials(credentials);
        return this;
    }
    signIn(code) {
        return __awaiter(this, void 0, void 0, function* () {
            // 認証情報を取得できればログイン成功
            const credentials = yield this.authClient.getToken(code, process.env.API_CODE_VERIFIER);
            debug('credentials published', credentials);
            if (credentials.access_token === undefined) {
                throw new Error('Access token is required for credentials.');
            }
            if (credentials.refresh_token === undefined) {
                throw new Error('Refresh token is required for credentials.');
            }
            // ログイン状態を保持
            const results = yield redisClient.multi()
                .set(`line-assistant.credentials.${this.userId}`, JSON.stringify(credentials))
                .expire(`line-assistant.credentials.${this.userId}`, EXPIRES_IN_SECONDS, debug)
                .exec();
            debug('results:', results);
            // rekognitionコレクション作成
            yield new Promise((resolve, reject) => {
                rekognition.createCollection({
                    CollectionId: this.rekognitionCollectionId
                }, (err, __) => __awaiter(this, void 0, void 0, function* () {
                    if (err instanceof Error) {
                        // すでに作成済であればok
                        if (err.code === 'ResourceAlreadyExistsException') {
                            resolve();
                        }
                        else {
                            reject(err);
                        }
                    }
                    else {
                        resolve();
                    }
                }));
            });
            // リフレッシュトークンを保管
            yield redisClient.multi()
                .set(`line-assistant.refreshToken.${this.userId}`, credentials.refresh_token)
                .expire(`line-assistant.refreshToken.${this.userId}`, REFRESH_TOKEN_EXPIRES_IN_SECONDS, debug)
                .exec();
            debug('refresh token saved.');
            this.setCredentials(Object.assign({}, credentials, { access_token: credentials.access_token }));
            return this;
        });
    }
    signInForcibly(credentials) {
        return __awaiter(this, void 0, void 0, function* () {
            // ログイン状態を保持
            const results = yield redisClient.multi()
                .set(`line-assistant.credentials.${this.userId}`, JSON.stringify(credentials))
                .expire(`line-assistant.credentials.${this.userId}`, EXPIRES_IN_SECONDS, debug)
                .exec();
            debug('results:', results);
            this.setCredentials(Object.assign({}, credentials, { access_token: credentials.access_token }));
            return this;
        });
    }
    logout() {
        return __awaiter(this, void 0, void 0, function* () {
            yield redisClient.del(`line-assistant.credentials.${this.userId}`);
        });
    }
    saveCallbackState(state) {
        return __awaiter(this, void 0, void 0, function* () {
            yield redisClient.multi()
                .set(`line-assistant.callbackState.${this.userId}`, state)
                .expire(`line-assistant.callbackState.${this.userId}`, EXPIRES_IN_SECONDS, debug)
                .exec();
        });
    }
    findCallbackState() {
        return __awaiter(this, void 0, void 0, function* () {
            return redisClient.get(`line-assistant.callbackState.${this.userId}`).then((value) => {
                return (value !== null) ? JSON.parse(value) : null;
            });
        });
    }
    deleteCallbackState() {
        return __awaiter(this, void 0, void 0, function* () {
            yield redisClient.del(`line-assistant.callbackState.${this.userId}`);
        });
    }
    saveMFAPass(pass, postEvent) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                // JWT作成
                const payload = { events: [postEvent] };
                jwt.sign(payload, SECRET, { expiresIn: POST_EVENT_TOKEN_EXPIRES_IN_SECONDS }, (jwtErr, token) => __awaiter(this, void 0, void 0, function* () {
                    if (jwtErr instanceof Error) {
                        reject(jwtErr);
                    }
                    else {
                        const key = `line-assistant.postEvent.${this.userId}.${pass}`;
                        const results = yield redisClient.multi()
                            .set(key, token)
                            .expire(key, POST_EVENT_TOKEN_EXPIRES_IN_SECONDS, debug)
                            .exec();
                        debug('results:', results);
                        resolve(pass);
                    }
                }));
            });
        });
    }
    verifyMFAPass(pass) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                const key = `line-assistant.postEvent.${this.userId}.${pass}`;
                const token = yield redisClient.get(key);
                if (token === null) {
                    resolve(null);
                    return;
                }
                jwt.verify(token, SECRET, (jwtErr, decoded) => {
                    if (jwtErr instanceof Error) {
                        reject(jwtErr);
                    }
                    else {
                        resolve(decoded);
                    }
                });
            }));
        });
    }
    deleteMFAPass(pass) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = `line-assistant.postEvent.${this.userId}.${pass}`;
            yield redisClient.del(key);
        });
    }
    /**
     * 顔画像を検証する
     * @param source 顔画像buffer
     */
    verifyFace(source) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                rekognition.searchFacesByImage({
                    CollectionId: this.rekognitionCollectionId,
                    FaceMatchThreshold: 0,
                    // FaceMatchThreshold: FACE_MATCH_THRESHOLD,
                    MaxFaces: 5,
                    Image: {
                        Bytes: source
                    }
                }, (err, data) => {
                    if (err instanceof Error) {
                        reject(err);
                    }
                    else {
                        resolve(data);
                    }
                });
            });
        });
    }
    /**
     * 顔画像を登録する
     * @param source 顔画像buffer
     */
    indexFace(source) {
        return __awaiter(this, void 0, void 0, function* () {
            yield new Promise((resolve, reject) => {
                rekognition.indexFaces({
                    CollectionId: this.rekognitionCollectionId,
                    Image: {
                        Bytes: source
                    },
                    DetectionAttributes: ['ALL']
                    // ExternalImageId: 'STRING_VALUE'
                }, (err, __) => {
                    if (err instanceof Error) {
                        reject(err);
                    }
                    else {
                        debug('face indexed.');
                        resolve();
                    }
                });
            });
        });
    }
    /**
     * 登録済顔画像を検索する
     */
    searchFaces() {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                rekognition.listFaces({
                    CollectionId: this.rekognitionCollectionId
                }, (err, data) => {
                    if (err instanceof Error) {
                        // コレクション未作成であれば空配列を返す
                        if (err.code === 'ResourceNotFoundException') {
                            resolve([]);
                        }
                        else {
                            reject(err);
                        }
                    }
                    else {
                        const faces = (data.Faces !== undefined) ? data.Faces : [];
                        resolve(faces);
                    }
                });
            });
        });
    }
}
exports.default = User;
